helm: consolidate IPSec and Wireguard encryption options

jibi opened the pull request:
This commit introduces a new helm option, encryption.type, which can be
used to specifies what method should be used for encryption (either
ipsec or wireguard).

The goal here is to add support for wireguard without introducing any
breaking change in the helm charts.

Fixes: #15483

gandro commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -417,6 +417,9 @@ encryption:
+  type: ipsec
+
   # -- Name of the key file inside the Kubernetes secret configured via secretName.
   keyFile: keys
```
minor nit: Can we add a comment or something to this and the following ipsec-specific values that it only applies to ipsec?

jibi replied:
Right, sure!

gandro replied:
Looks good, thanks! :pray:

jibi replied:
(I also updated the `cilium-agent-daemonset.yaml` template as just found I missed it)

gandro reviewed (approved):
Thanks for taking care of this! One non-blocking nit.

gandro reviewed (changes requested):
Ah, good catch with the daemonset. I think you missed one though:
https://github.com/cilium/cilium/blob/ff183bdcb54526123f7fe60bbe3f35f784350ec8/install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml#L329

jibi committed:
helm: consolidate IPSec and Wireguard encryption options

This commit introduces a new helm option, encryption.type, which can be
used to specifies what method should be used for encryption (either
ipsec or wireguard).

The goal here is to add support for wireguard without introducing any
breaking change in the helm charts.

Fixes: #15483

Co-authored-by: Youssef Azrak <yazrak.tech@gmail.com>
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

gandro reviewed (approved):
:100:

brb reviewed (approved):
LGTM, thanks! One nit: I think we shall acredit the previous PR author in your commit msg too.

jibi committed:
docs: update Wireguard Getting Started Guide

This commit updates the Wireguard GSG to reflect the new Helm options
introduced to enable it.

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

nbusseneau commented on install/kubernetes/cilium/README.md:
```diff
@@ -122,10 +122,11 @@ contributors across the globe, there is almost always someone available to help.
-| encryption.mountPath | string | `"/etc/ipsec"` | Path to mount the secret inside the Cilium pod. |
-| encryption.nodeEncryption | bool | `false` | Enable encryption for pure node to node traffic. |
-| encryption.secretName | string | `"cilium-ipsec-keys"` | Name of the Kubernetes secret containing the encryption keys. |
+| encryption.keyFile | string | `"keys"` | Name of the key file inside the Kubernetes secret configured via secretName. This option is only effective when encryption.type is set to ipsec. |
+| encryption.mountPath | string | `"/etc/ipsec"` | Path to mount the secret inside the Cilium pod. This option is only effective when encryption.type is set to ipsec. |
+| encryption.nodeEncryption | bool | `false` | Enable encryption for pure node to node traffic. This option is only effective when encryption.type is set to ipsec. |
+| encryption.secretName | string | `"cilium-ipsec-keys"` | Name of the Kubernetes secret containing the encryption keys. This option is only effective when encryption.type is set to ipsec. |
+| encryption.type | string | `"ipsec"` | Encryption method. Can be either ipsec or wireguard. |
```
Since all these encryption options are for `ipsec` rather than for both `ipsec` and `wireguard`, would it make sense to instead have an `ipsec` block for holding them, rather than bundling them in `encryption`?

E.g. either `encryption.ipsec.keyFile` or `ipsec.keyFile`.

We could then mark the existing `encryption.keyFile` and friends as deprecated (so as not to break anything for now) and eventually remove them in a few releases.

Reasoning: I'm not a fan of keeping a set of options that only work if another option at the same level is set to a specific value. I feel like this does not convey the intent properly, because there is no explicit link between the two, except from the documentation / diving into the chart itself. Hence why I think it makes more sense for specific options to be moved to a dedicated lower-level.

jibi replied:
Yeah, I think that's a good shout :+1: 

What I would do is:
* duplicate `keyFile`, `mountPath`, `nodeEncryption` and `secretName` under `encryption.ipsec`
* mark the options under `encryption` as deprecated
* give the deprecated options precedence over the ones under `encryption.ipsec` (in case they are both set)

@gandro @brb thoughts?

gandro replied:
I really don't have a preference as to whether or not we should do this this. I leave that discussion up to people with more experience with Helm best practices.

However, if we do it, then I think we should keep `nodeEncryption` generic. Even if it is not supported by the Wireguard-based encryption right now, it will likely become supported by Cilium 1.11.

The `keyFile`/`mountPath`/`secretName` however can probably safely be moved to a `ipsec` section. Even if we add support in the charts to inject a custom Wireguard private key via K8s secret in the future, the private keys for Wireguard needs to be specified per-node (where ipsec has a single key valid for all nodes) and thus the Helm values structure for that would look different as they will for ipsec.

jibi replied:
@nbusseneau I would then propose to merge this PR as it is since this is a release blocker, and add the new flags/deprecate the old ones in a separate PR, what do you think?

nbusseneau replied:
Sure!

nbusseneau reviewed (commented):
LGTM as-is, just one reserve below.
